### PR TITLE
feat: terramate list --deployment-status=<status> and --drift-status=<status>

### PR DESCRIPTION
**Pull request by i4ki**
## What this PR does / why we need it:

Introduce `terramate [list,run] --[deployment,drift]-status=<status>` flags.

## Which issue(s) this PR fixes:
none

## Does this PR introduce a user-facing change?
```
yes, adds a new feature.
```

### CODE REVIEW

File: cloud/drift/status.go [hunk 1]
```diff
@@ -16,24 +16,44 @@ import (
...
 )
 
 // NewStatus creates a new stack drift status from a string.
-func NewStatus(str string) (Status, error) {
+func NewStatus(str string) Status {
 	var s Status
 	err := s.UnmarshalJSON([]byte(strconv.Quote(str)))
-	return s, err
+	if err != nil {
+		panic(errors.E(errors.ErrInternal, "Terramate must not fail when parsing drift statuses"))
```
Comment (snakster) [inline cloud/drift/status.go:54]: There are 3 implementations for this in deployment, stack and drift, and only two of them panic, the one for stack just continues. Maybe worth aligning them?

File: cloud/drift/status.go [hunk 2]
```diff
@@ -85,6 +105,50 @@ func (s Status) String() string {
...
+func NewStatusFilter(str string) (FilterStatus, error) {
+	if str == "unhealthy" {
+		return UnhealthyFilter, nil
+	} else if str == "healthy" {
+		return HealthyFilter, nil
+	}
+
+	s := NewStatus(str)
+	if s == Unrecognized {
+		return FilterStatus(0), errors.E("unrecognized drift status filter: %s", str)
```
Comment (snakster) [inline cloud/drift/status.go:128]: The parsing error for the different filter flags is slightly inconsistent in it's wording:
```
% terramate list --drift-status=foobar
Error: failed to parse drift filter
> unrecognized drift status filter: foobar

% terramate list --status=foobar       
Error: unrecognized stack filter
> unrecognized status filter: foobar

% terramate list --deployment-status=foobar
Error: failed to parse deployment filter
> unrecognized deployment status filter: foobar
```

File: cloud/drift/status.go [hunk 3]
```diff
@@ -16,24 +16,44 @@ import (
 const (
 	// ErrInvalidStatus represents an invalid drift status error.
 	ErrInvalidStatus errors.Kind = "invalid drift status"
+	// ErrInvalidFilterStatus represents an invalid drift status filter error.
+	ErrInvalidFilterStatus errors.Kind = "invalid filter stack status"
```
Comment (snakster) [inline cloud/drift/status.go:20]:
```suggestion
	ErrInvalidFilterStatus errors.Kind = "invalid filter drift status"
```

File: cloud/deployment/status.go [hunk 4]
```diff
@@ -16,26 +16,50 @@ import (
 const (
 	// ErrInvalidStatus represents an invalid deployment status error.
 	ErrInvalidStatus errors.Kind = "invalid deployment status"
+
+	// ErrInvalidFilterStatus represents an invalid deployment status filter error.
+	ErrInvalidFilterStatus errors.Kind = "invalid filter stack status"
```
Comment (snakster) [inline cloud/deployment/status.go:21]: Intended or forgotten to change?
```suggestion
	ErrInvalidFilterStatus errors.Kind = "invalid filter deployment status"
```

File: cloud/types.go [hunk 5]
```diff
@@ -698,3 +705,16 @@ func (c LogChannel) String() string {
 	}
 	return "unknown"
 }
+
+// NoStackFilter returns a StatusFilters with no filter.
+func NoStackFilter() StatusFilters {
+	return StatusFilters{
+		StackStatus:      stack.NoFilter,
+		DeploymentStatus: deployment.NoFilter,
```
Comment (snakster) [inline cloud/types.go:713]: What about `DriftStatus`?

File: cmd/terramate/cli/cli.go [hunk 6]
```diff
@@ -325,6 +322,14 @@ type safeguards struct {
 	reEnabled bool
 }
 
+type cloudFilterFlags struct {
+	ExperimentalStatus string `hidden:"" help:"Filter by status (Deprecated)"`
+	CloudStatus        string `hidden:""`
+	Status             string `help:"Filter by Terramate Cloud status of the stack."`
+	DeploymentStatus   string `help:"Filter by Terramate Cloud Deployment Status of the stack"`
```
Comment (snakster) [inline cmd/terramate/cli/cli.go:329]:
```suggestion
	DeploymentStatus   string `help:"Filter by Terramate Cloud deployment status of the stack"`
```

File: cmd/terramate/cli/cli.go [hunk 7]
```diff
@@ -325,6 +322,14 @@ type safeguards struct {
...
+	DriftStatus        string `help:"Filter by Terramate Cloud Drift Status of the stack"`
```
Comment (snakster) [inline cmd/terramate/cli/cli.go:330]:
```suggestion
	DriftStatus        string `help:"Filter by Terramate Cloud drift status of the stack"`
```

File: cloud/types.go [hunk 8]
```diff
@@ -698,3 +705,16 @@ func (c LogChannel) String() string {
 	}
 	return "unknown"
 }
+
+// NoStackFilter returns a StatusFilters with no filter.
+func NoStackFilter() StatusFilters {
```
Comment (snakster) [inline cloud/types.go:710]: Maybe name this NoStatusFilters instead?

File: cmd/terramate/cli/cli.go [hunk 9]
```diff
@@ -1033,7 +1040,7 @@ func (c *cli) triggerStack(basePath string) {
 		stacks = append(stacks, st.Sortable())
 	} else {
 		var err error
-		stacksReport, err := c.listStacks(false, cloudstack.AnyTarget, cloudstack.NoFilter)
+		stacksReport, err := c.listStacks(false, cloudstack.AnyTarget, cloud.StatusFilters{})
```
Comment (snakster) [inline cmd/terramate/cli/cli.go:1043]: Should we use the no filters function here?

File: cmd/terramate/cli/cli.go [hunk 10]
```diff
@@ -1743,19 +1758,41 @@ func (c *cli) printStacksList(allStacks []stack.Entry, why bool, runOrder bool)
...
-		if status.Is(cloudstack.Unrecognized) {
-			fatalf("unrecognized stack filter: %s", strStatus)
-		}
+func parseStatusFilter(filterStr string) cloudstack.FilterStatus {
+	if filterStr == "" {
+		return cloudstack.NoFilter
+	}
+	filter, err := cloudstack.NewStatusFilter(filterStr)
+	if err != nil {
+		fatalWithDetails(err, "unrecognized stack filter")
```
Comment (snakster) [inline cmd/terramate/cli/cli.go:1767]: I see  this is where my earlier  comment regarding the different error wording is coming from. I think they should all be aligned.

**Review by snakster — COMMENTED** [review]
Nice work! Left a few small suggestions for improvement.

File: cloud/deployment/status.go (hunk 4 above)
Comment (i4ki) [reply cloud/deployment/status.go:21]: so much copy paste, sorry.

File: cloud/drift/status.go (hunk 1 above)
Comment (i4ki) [reply cloud/drift/status.go:54]: it should never fail unless JSON encoding has issues encoding a quoted string.

Comment (i4ki) [reply cloud/drift/status.go:54]: I will remove this panic in this and other cases

**Review by snakster — DISMISSED** [review]
👍

**Review by snakster — APPROVED** [review]

**Pull request merged**